### PR TITLE
Add orderBy clause support to aggregate function builder

### PR DESCRIPTION
**Pull request by ivashog**
#780

see sql syntax on PostgreSQL example: https://www.postgresql.org/docs/16/sql-expressions.html#SYNTAX-AGGREGATES

### CODE REVIEW

**Comment by igalklebanov** [comment]
Hey 👋 

Thanks! 💪 

This is also support in [MySQL](https://dev.mysql.com/doc/refman/8.3/en/aggregate-functions.html#function_group-concat) and [SQLite](https://www.sqlite.org/lang_aggfunc.html) for some aggregate functions. Let's add tests for those.

**Comment by igalklebanov** [comment]
> But it is better to implement it in a separate PR, what do you say?

Yeah, different clause, different PR.

File: test/node/src/aggregate-function.test.ts [hunk 1]
```diff
@@ -1106,6 +1107,56 @@ for (const dialect of DIALECTS) {
 
       await query.execute()
     })
+
+    describe(`should execute order-sensitive aggregate functions`, () => {
+      // @todo add test for mssql when implement `WITHIN GROUP ( order_by_clause )` clause
```
Comment (igalklebanov) [inline test/node/src/aggregate-function.test.ts:1112]:
```suggestion
```

File: test/node/src/aggregate-function.test.ts [hunk 2]
```diff
@@ -1106,6 +1107,56 @@ for (const dialect of DIALECTS) {
...
+      if (dialect !== 'mssql') {
```
Comment (igalklebanov) [inline test/node/src/aggregate-function.test.ts:1113]: Now that we have many dialects, it's best to be explicit.

```suggestion
      if (dialect === 'postgres' || dialect === 'mysql' || dialect === 'sqlite') {
```

File: test/node/src/aggregate-function.test.ts [hunk 3]
```diff
@@ -1106,6 +1107,56 @@ for (const dialect of DIALECTS) {
...
+        const isMySql = dialect === 'mysql'
+        const funcName = isMySql ? 'group_concat' : 'string_agg'
+        const funcArgs: Array<ReferenceExpression<Database, 'person'>> = [
+          'first_name',
+        ]
+        if (!isMySql) funcArgs.push(sql.lit(','))
```
Comment (igalklebanov) [inline test/node/src/aggregate-function.test.ts:1119]: nitpicky

```suggestion
        if (!isMySql) {
          funcArgs.push(sql.lit(','))
        }
```

File: test/node/src/aggregate-function.test.ts [hunk 4]
```diff
@@ -1106,6 +1107,56 @@ for (const dialect of DIALECTS) {
...
+            postgres: {
+              sql: [
+                `select ${funcName}("first_name", ',' order by "first_name" desc) as "first_names"`,
+                `from "person"`,
+              ],
+              parameters: [],
+            },
+            mysql: {
+              sql: [
+                `select group_concat(\`first_name\` order by \`first_name\` desc) as \`first_names\``,
```
Comment (igalklebanov) [inline test/node/src/aggregate-function.test.ts:1141]:
```suggestion
                `select ${funcName}(\`first_name\` order by \`first_name\` desc) as \`first_names\``,
```

File: test/node/src/aggregate-function.test.ts [hunk 5]
```diff
@@ -1106,6 +1107,56 @@ for (const dialect of DIALECTS) {
...
+                `from \`person\``,
+              ],
+              parameters: [],
+            },
+            mssql: NOT_SUPPORTED,
+            sqlite: {
+              sql: [
+                `select string_agg("first_name", ',' order by "first_name" desc) as "first_names"`,
```
Comment (igalklebanov) [inline test/node/src/aggregate-function.test.ts:1149]:
```suggestion
                `select ${funcName}("first_name", ',' order by "first_name" desc) as "first_names"`,
```

File: src/query-builder/aggregate-function-builder.ts [hunk 6]
```diff
@@ -95,6 +103,43 @@ export class AggregateFunctionBuilder<DB, TB extends keyof DB, O = unknown>
     })
   }
 
+  /**
+   * Adds an `order by` clause inside the aggregate function for specify correct result order
+   * ([see postgres docs](https://www.postgresql.org/docs/9.5/sql-expressions.html#SYNTAX-AGGREGATES))
```
Comment (igalklebanov) [inline src/query-builder/aggregate-function-builder.ts:108]: This is more than enough. "Bring Your Own SQL Knowledge".

```suggestion
   * Adds an `order by` clause inside the aggregate function.
```

File: test/node/src/aggregate-function.test.ts (hunk 1 above)
Comment (igalklebanov) [reply test/node/src/aggregate-function.test.ts:1112]: We should open an issue for this, and not leave a comment in the code.

**Review by igalklebanov — COMMENTED** [review]
Looks good and functional overall! 💪 

Left a few minor suggestions, once addressed this is good to go! 🚀

**Comment by ivashog** [comment]
> Looks good and functional overall! 💪
> 
> Left a few minor suggestions, once addressed this is good to go! 🚀

@igalklebanov, thanks for review!

File: test/typings/test-d/aggregate-function.test-d.ts [hunk 7]
```diff
@@ -378,6 +378,23 @@ async function testSelectWithDistinct(db: Kysely<Database>) {
...
+        .agg<string[]>('array_agg', ['pet.name'])
+        .orderBy('pet.name')
+        .as('person_pet_names'),
+      eb.fn.jsonAgg('pet').orderBy('pet.id').as('person_pets'),
+    ])
+    .executeTakeFirstOrThrow()
+
+  expectAssignable<string[]>(result.person_pet_names)
+  expectAssignable<Selectable<Pet>[]>(result.person_pets)
+}
```
Comment (igalklebanov) [inline test/typings/test-d/aggregate-function.test-d.ts:396]: I don't think this adds any value. Let's remove this.

**Review by igalklebanov — APPROVED** [review]
LGTM! 🚀

File: test/typings/test-d/aggregate-function.test-d.ts (hunk 7 above)
Comment (ivashog) [reply test/typings/test-d/aggregate-function.test-d.ts:396]: I agree

**Comment by juusaw** [comment]
Any plans to release this feature? It would be really useful for us.

**Pull request merged**